File: tests/usecase/role/list_test.rs
use std::sync::Arc;

use async_trait::async_trait;

use go_clean_template::application::usecase::role::{ListRolesInput, ListRolesUseCase, ListRolesUseCaseImpl};
use go_clean_template::application::usecase::user::RoleRepository;
use go_clean_template::domain;

struct MockRoleRepo {
    roles: Vec<domain::Role>,
}

#[async_trait]
impl RoleRepository for MockRoleRepo {
    async fn find_by_id(&self, _id: &str) -> Result<domain::Role, domain::Error> {
        Err(domain::Error::NotFound)
    }
    async fn find_all(&self, offset: usize, limit: usize) -> Result<(Vec<domain::Role>, i64), domain::Error> {
        let total = self.roles.len() as i64;
        let slice = self.roles.iter().skip(offset).take(limit).cloned().collect();
        Ok((slice, total))
    }
    async fn find_by_name(&self, _name: &str) -> Result<domain::Role, domain::Error> {
        Err(domain::Error::NotFound)
    }
    async fn create(&self, _role: &domain::Role) -> Result<(), domain::Error> { Ok(()) }
    async fn update(&self, _role: &domain::Role) -> Result<(), domain::Error> { Ok(()) }
    async fn delete(&self, _id: &str) -> Result<(), domain::Error> { Ok(()) }
}

#[tokio::test]
async fn test_list_roles_success() {
    let repo = MockRoleRepo {
        roles: vec![
            domain::Role::new("admin".into(), "Admin".into()).unwrap(),
            domain::Role::new("user".into(), "User".into()).unwrap(),
        ],
    };

    let uc = ListRolesUseCaseImpl::new(Arc::new(repo));
    let result = uc.execute(ListRolesInput { page: 1, page_size: 20 }).await;
    assert!(result.is_ok());
    assert_eq!(result.unwrap().roles.len(), 2);
}

#[tokio::test]
async fn test_list_roles_empty() {
    let repo = MockRoleRepo { roles: vec![] };
    let uc = ListRolesUseCaseImpl::new(Arc::new(repo));
    let result = uc.execute(ListRolesInput { page: 1, page_size: 20 }).await;
    assert!(result.is_ok());
    assert!(result.unwrap().roles.is_empty());
}

File: tests/application/usecase/role/list_test.go
package role_test

import (
	"context"
	"testing"

	"github.com/stretchr/testify/assert"

	roleusecase "go-clean-template/internal/application/usecase/role"
	"go-clean-template/internal/domain"
)

func TestRoleGetAll_Success(t *testing.T) {
	mockRepo := new(MockRoleRepo)
	uc := roleusecase.NewListRolesUseCase(mockRepo)

	roles := []*domain.Role{
		{ID: "1", Name: "admin", Description: "Admin"},
		{ID: "2", Name: "user", Description: "User"},
	}
	mockRepo.On("FindAll").Return(roles, nil)

	resp, err := uc.Execute(context.Background(), &roleusecase.ListRolesInput{Page: 1, PageSize: 20})

	assert.NoError(t, err)
	assert.Equal(t, int64(2), resp.Total)
	assert.Len(t, resp.Roles, 2)
	mockRepo.AssertExpectations(t)
}

func TestRoleGetAll_Empty(t *testing.T) {
	mockRepo := new(MockRoleRepo)
	uc := roleusecase.NewListRolesUseCase(mockRepo)

	mockRepo.On("FindAll").Return([]*domain.Role{}, nil)

	resp, err := uc.Execute(context.Background(), &roleusecase.ListRolesInput{Page: 1, PageSize: 20})

	assert.NoError(t, err)
	assert.Equal(t, int64(0), resp.Total)
	assert.Empty(t, resp.Roles)
	mockRepo.AssertExpectations(t)
}
